### ISSUE TITLE
Softmax needs arbitrary axis support

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2398]
## Description

The softmax activation function only supports a specific axis and needs to support arbitrary axis specification.

## Location

- File: `shared/core/activation.mojo`
- Line: Softmax implementation

## TODO/FIXME Text

```
# TODO: Support arbitrary axis with proper reduction
```

## Current Behavior

Softmax likely only works on a fixed axis (e.g., last axis or axis=1).

## Expected Behavior

```mojo
# Should support any axis
var probs = softmax(logits, axis=1)   # Along classes (common for classification)
var probs = softmax(logits, axis=-1)  # Last axis
var probs = softmax(logits, axis=0)   # Along batch (less common but valid)
```

## Acceptance Criteria

- [ ] Add `axis` parameter to softmax function
- [ ] Support negative axis indexing
- [ ] Proper reduction along specified axis
- [ ] Numerical stability (subtract max before exp)
- [ ] Preserve other dimensions correctly
- [ ] Tests for various axis values and tensor shapes
- [ ] Match NumPy/PyTorch softmax behavior

## Impact

Limited axis support restricts model architectures that may need softmax along different dimensions.
